app/Main.hs

```diff
@@ -146,23 +162,39 @@ vaultRetryPolicy opts = Retry.fullJitterBackoff (unMilliSeconds (oRetryBaseDelay
 
 main :: IO ()
 main = do
-  env <- getEnvironment
-  opts <- execParser (optionsInfo env)
+  localEnvVars <- getEnvironment
+  cliOptions <- execParser (optionsInfo localEnvVars)
+  httpManager <- getHttpManager cliOptions
+
+  let context = Context { cLocalEnvVars = localEnvVars
+                        , cCliOptions = cliOptions
+                        , cHttpManager = httpManager
+                        }
 
-  eres <- runExceptT $ runReaderT vaultEnv (env, opts)
-  case eres of
+  runExceptT (vaultEnv context) >>= \case
     Left err -> hPutStrLn stderr (vaultErrorLogMessage err)
-    Right newEnv -> runCommand opts newEnv
-
-vaultEnv :: ReaderT ExecCtx (ExceptT VaultError IO) [EnvVar]
-vaultEnv = do
-  secretFile <- asks (oSecretFile . snd)
-  secrets <- readSecretList secretFile
-  opts <- asks snd
-  secretEnv <- requestSecrets opts secrets
-  localEnv <- asks fst
-  inheritEnvOff <- asks (oInheritEnvOff . snd)
-  checkNoDuplicates (buildEnv localEnv secretEnv inheritEnvOff)
+    Right newEnv -> runCommand cliOptions newEnv
+
+-- | This function returns either a manager for plain HTTP or
+-- for HTTPS connections. If TLS is wanted, we also check if the
+-- user specified an option to disable the certificate check.
+getHttpManager :: Options -> IO Manager
+getHttpManager opts = newManager managerSettings
+  where
+    managerSettings = if oConnectHttp opts
+                      then defaultManagerSettings
+                      else mkManagerSettings tlsSettings Nothing
+    tlsSettings = TLSSettingsSimple
+                { settingDisableCertificateValidation = oNoValidateCerts opts
+                , settingDisableSession = False
+                , settingUseServerName = True
+                }
+
+vaultEnv :: Context -> (ExceptT VaultError IO) [EnvVar]
```

Comment:
Docstring please.

Reply:
I would also drop the parentheses around `(ExceptT VaultError IO)`. To me they are more confusing than helpful.

Reply:
Done. They were left over from when the typesignature was `vaultEnv :: ReaderT ExecCtx (ExceptT VaultError IO) [EnvVar]`